Fix remaining blackmail vulnerabilities

stejbac opened the pull request:
<!-- 
- make yourself familiar with the CONTRIBUTING.md if you have not already (https://github.com/bisq-network/bisq/blob/master/CONTRIBUTING.md)
- make sure you follow our [coding style guidelines][https://github.com/bisq-network/style/issues)
- pick a descriptive title
- provide some meaningful PR description below
- create the PR
- in case you receive a "Change request" and/or a NACK, please react within 30 days. If not, we will close your PR and it can not be up for compensation.
- After addressing the change request, __please re-request a review!__ Otherwise we might miss your PR as we tend to only look at pull requests tagged with a "review required".
-->

This PR attempts to take advantage of the SegWit trade protocol upgrade (hard fork), to fix all the remaining cases where a buyer or seller can publish a deposit tx without their peer having a valid delayed payout tx (at least in the fully SegWit case, when taking an offer created in v1.5.0+). When taking pre-1.5.0 offers, the deposit tx inputs will be of mixed type, which makes it malleable and thus impossible to completely prevent either trader from altering its ID and thus rendering their peer's delayed payout tx invalid. In that case, the buyer should still fully validate their delayed payout tx anyway, against the agreed deposit tx, so that if the latter changes before confirming, the trade hopefully fails before the buyer makes payment. The PR additionally fixes a bug where (it appears) the buyer fails to check the signature of the final delayed payout tx.

These changes are intended to prevent blackmail attacks (or reduce the risk of them in not-fully-SegWit cases where the deposit tx is malleable).

stejbac committed:
Fix missing segwit case when sanitising preparedDepositTx

Make sure witness data is stripped from the seller's prepared deposit
tx, in addition to ScriptSig data, to prevent the buyer from being able
to publish it prematurely (before having signed the delayed payout tx).

stejbac committed:
Send seller's delayedPayoutTx signature to peer ASAP

Include a new 'delayedPayoutTxSellerSignature' field with the prepared
delayed payout tx sent to the buyer, in DelayedPayoutTxSignatureRequest.
This will allow the buyer to compute the final, signed delayedPayoutTx
as early as possible and withhold their deposit tx witness from the
seller until they know they have a valid delayedPayoutTx, preventing its
premature publishing in the fully segwit case. (To be done in a later
commit - for now just save the seller's delayedPayoutTx signature.)

As part of this, run the SellerSignsDelayedPayoutTx trade task at an
earlier step (just after payout tx creation) to make its signature
available to the seller ASAP. Also rename 'delayedPayoutTxSignature' to
'delayedPayoutTxBuyerSignature' in DelayedPayoutTxSignatureResponse.

stejbac committed:
Add new BuyerFinalizesDelayedPayoutTx task

Improve validation of the buyer's delayed payout tx (both before & after
they get the final DepositTxAndDelayedPayoutTxMessage from the peer), by
finalising it independently of the seller. This is now possible since
their 2-of-2 signature is included in the DelayedPayoutSignatureRequest.
Check that the final delayedPayoutTx received from the seller matches it
byte-for-byte (which actually makes its receipt redundant now).

This also fixes an apparent security bug, where the final validation of
the delayedPayoutTx appears to skip any kind of signature check (only a
deposit tx hash check, which is still necessary).

Finally, optimistically check the deposit tx against the input of the
prepared delayedPayoutTx received from the seller, in the case that the
former is non-malleable (that is, the fully segwit case) and thus has a
stable ID given by the hash of the buyer's preparedDepositTx.

stejbac committed:
Withhold witnesses in buyer->seller depositTx data, until last step

Strip all input witnesses from the depositTx message fields sent from
the buyer, until the last (DelayedPayoutTxSignatureResponse) message is
sent, where they can be bundled in as an extra field. Since the witness
data doesn't affect the final deposit tx id, the seller does not need to
know it until actually publishing the tx.

In the (fully) segwit case, this allows the buyer to prevent the seller
from publishing the deposit tx until the buyer has a valid, fully signed
delayedPayoutTx. Provide the final witness data in an extra 'depositTx'
field in DelayedPayoutTxSignatureResponse, which the seller can merge
with his depositTx witness block (for his own input signatures).

stejbac committed:
Prevent takers from using non-segwit deposit tx inputs

Disallow non-P2WH depositTx inputs from the taker, while continuing to
allow them from the maker, so that offers created pre-v1.5.0 can still
be taken. (After some time, those inputs could be disallowed too.)

This is mainly to prevent mass blackmail attacks, where more victims'
money could be locked up than the DAO could possibly compensate them all
for. (This is probably only an attractive attack for a buyer anyway, at
least with the earlier commits.)

chimp1984 commented on core/src/main/java/bisq/core/trade/protocol/tasks/buyer_as_maker/BuyerAsMakerSendsInputsForDepositTxResponse.java:
```diff
@@ -32,6 +34,9 @@ public BuyerAsMakerSendsInputsForDepositTxResponse(TaskRunner<Trade> taskHandler
+        Transaction preparedDepositTx = processModel.getBtcWalletService().getTxFromSerializedTx(processModel.getPreparedDepositTx());
+        // Remove witnesses from preparedDepositTx, so that the seller can still compute the final
+        // tx id, but cannot publish it before providing the buyer with a signed delayed payout tx.
+        return preparedDepositTx.bitcoinSerialize(false);
```
The Transaction.bitcoinSerializeToStream method does use that flag not only for adding segwit data but also for writing a flag: 
```
 // marker, flag
        if (useSegwit) {
            stream.write(0);
            stream.write(1);
        }
```

I don't know which function this flag has, but maybe its more safe to remove the segwit data in a different way. Maybe @oscarguindzberg can add his input here...

oscarguindzberg replied:
the function of the flag is "tx serialzed using the new segwit format".
bitcoinSerialize(false) seems fine if you want to avoid sending the signatures.

sqrrm commented:
I like this PR and it should be included in v1.5.0. From what I can see it looks reasonable but I'm not familiar enough with all the bitcoinj details.

stejbac commented:
I just spotted that I accidentally used the non-segwit version of `Script.correctlySpends`, when checking the p2wsh signature of the delayed payout tx (during `BuyerFinalizesDelayedPayoutTx`), on line 774 of `TradeWalletService`:

> input.getScriptSig().correctlySpends(delayedPayoutTx, 0, scriptPubKey, Script.ALL_VERIFY_FLAGS);

This erroneously skips the check, as the ScriptSig is empty. I need to additionally pass the witness and locked up trade amount. I'll post another commit to fix this shortly.

chimp1984 reviewed (approved):
<strike>utACK</strike>

Ah read the last comment just now. So I will wait for the open fix for another review...

chimp1984 commented:
@oscarguindzberg Could you please review that PR as well?

stejbac committed:
Fix faulty signature check in finalizeDelayedPayoutTx

Make sure to use the segwit version of Script.correctlySpends in
TradeWalletService.finalizeDelayedPayoutTx, which requires the input
value and witness to be passed explicitly (as the latter holds the
actual signature). This was causing BuyerFinalizesDelayedPayoutTx to
fail to do any kind of signature check.

Also refactor the method slightly and remove a redundant call to
WalletService.checkScriptSig (which does the same thing as
TransactionInput.verify) in the branch used by the seller.

chimp1984 reviewed (approved):
utACK

sqrrm commented:
@stejbac The build is failing

@oscarguindzberg I will wait for you ack before merging this

stejbac commented:
@sqrrm I'm not sure why the build is failing (or the Codacy checks). The Travis build log of one job, which ran 9 hours ago, has the following error:

> Could not GET 'https://repo.maven.apache.org/maven2/org/openjfx/javafx-base/11.0.2/javafx-base-11.0.2.jar'.

The other job (apparently run around the same time) has failures in `OpenOfferManagerTest` (lines 45, 106) in the _:core:test_ task, but I'm fairly sure that's not caused by my changes.

ghubstan commented:
> ... I'm not sure why the build is failing (or the Codacy checks). The Travis build log of one job, which ran 9 hours ago, has the following error:
> 
> > Could not GET 'https://repo.maven.apache.org/maven2/org/openjfx/javafx-base/11.0.2/javafx-base-11.0.2.jar'.
> 
> The other job (apparently run around the same time) has failures in `OpenOfferManagerTest` (lines 45, 106) in the _:core:test_ task, but I'm fairly sure that's not caused by my changes.

Which jar travis cannot download at any given time is fairly random.  Closing and re-opening the PR will re-run the travis build, and the dependency download will probably work.    However, Codacy's `Seems like the release/v1.5.0 is not being analyzed. To review this pull request enable the branch in the repository settings.` is another issue that someone (@ripcurlx ?) needs to look at.

ripcurlx commented:
> > ... I'm not sure why the build is failing (or the Codacy checks). The Travis build log of one job, which ran 9 hours ago, has the following error:
> > > Could not GET 'https://repo.maven.apache.org/maven2/org/openjfx/javafx-base/11.0.2/javafx-base-11.0.2.jar'.
> > 
> > 
> > The other job (apparently run around the same time) has failures in `OpenOfferManagerTest` (lines 45, 106) in the _:core:test_ task, but I'm fairly sure that's not caused by my changes.
> 
> Which jar travis cannot download at any given time is fairly random. Closing and re-opening the PR will re-run the travis build, and the dependency download will probably work. However, Codacy's `Seems like the release/v1.5.0 is not being analyzed. To review this pull request enable the branch in the repository settings.` is another issue that someone (@ripcurlx ?) needs to look at.

Yes, please ignore the Codacy warning. It is just about the fact that it is not configured to run on this branch. I just manually triggered a travis re-build for this PR.

ripcurlx commented:
I just saw in Travis that there are two failing tests right now:
```
bisq.core.offer.OpenOfferManagerTest > testStartEditOfferForActiveOffer FAILED
    java.lang.RuntimeException at OpenOfferManagerTest.java:45
bisq.core.offer.OpenOfferManagerTest > testStartEditOfferForOfferThatIsCurrentlyEdited FAILED
    java.lang.RuntimeException at OpenOfferManagerTest.java:106
```

ripcurlx commented:
> I just saw in Travis that there are two failing tests right now:
> 
> ```
> bisq.core.offer.OpenOfferManagerTest > testStartEditOfferForActiveOffer FAILED
>     java.lang.RuntimeException at OpenOfferManagerTest.java:45
> bisq.core.offer.OpenOfferManagerTest > testStartEditOfferForOfferThatIsCurrentlyEdited FAILED
>     java.lang.RuntimeException at OpenOfferManagerTest.java:106
> ```

🤔 it does run locally without any problems.

stejbac commented:
I got the failing tests locally after I tried merging onto release/1.5.0, so the tip map be broken right now. It seems it can be fixed by adding tearDown() methods in the failing tests to call PersistenceManager.shutDown, which removes the particular persistence manager instance created in each test from a global map. (All but one tests in each affected test class fail, presumably depending in the order they run in.)

ripcurlx commented:
> I got the failing tests locally after I tried merging onto release/1.5.0, so the tip map be broken right now. It seems it can be fixed by adding tearDown() methods in the failing tests to call PersistenceManager.shutDown, which removes the particular persistence manager instance created in each test from a global map. (All but one tests in each affected test class fail, presumably depending in the order they run in.)

Yes, also https://github.com/bisq-network/bisq/pull/4816 is affected by this right now. Could you please push this commit to this branch to see if Travis is happy with it as well and I'll cherry-pick and push it to the other one as well?

stejbac commented:
It looks like it fixed the jdk10 job but not the jdk12 job, which I guess has an unrelated problem.

ripcurlx commented:
> It looks like it fixed the jdk10 job but not the jdk12 job, which I guess has an unrelated problem.

I just restarted the job. Travis is kind of unstable the last couple of days.

stejbac committed:
Add further validation checks for delayed payout tx

Do some extra sanity checks like tx.outputSum < tx.inputSum, to rule out
any edge cases where an invalid delayed payout tx might still arise.

ripcurlx commented:
@stejbac As I just merged #4816. Could you please resolve the conflict? Thanks!

stejbac commented:
@ripcurlx OK, I guess I should rebase onto release/1.5.0 to resolve the conflict.

chimp1984 reviewed (approved):
utACK
